### DISCUSSION TITLE
NATS Task Scheduler

### DISCUSSION THREAD

**Discussion by tamis-laan** [nats-io/nats-server#5733] [Ideas]
I'm in a situation where I need to trigger tasks based on time. NATS already provides a key value store and a object store and it would be great to also have a task scheduler. 

Task schedulers such as Celery make use of Redis or RabbitMQ as a broker but I would not like to host another key value store besides the NATS key value store.

There is already something called [GoCelery](https://github.com/gocelery/gocelery) that can use NATS as the broker but this is bound specifically to golang.

In essence it wouldn't need to be a "Task" scheduler but more like a message scheduler. The message would then trigger a worker that does the processing.

In fact this could be build using the `Nats-Delay` header for scheduling a message and storing the (stream, sequence number) in a key-value store so we can edit or delete the scheduled message at a later date. We also need to make sure the key value item is deleted after the message has been scheduled but this could be done with a TTL on the key-value pair.

**Comment by nickchomey** [comment]
I came across this package the other day, which is part of a broader toolset that is developed by one of the nats/synadia devs. Might be worth a look! 

https://github.com/choria-io/asyncjobs

**Reply by tamis-laan** [reply]
@nickchomey

Hey nick, thanks for the link. It does look similar to what I'm interested in however this would be great as a native feature and not just bound to golang. In my case I'm using python.
